src/Tokenizer/Tokens.php

```diff
@@ -363,6 +363,13 @@ public function offsetSet($index, $newval): void
             )));
         }
 
+        if (!$newval instanceof Token) {
+            Utils::triggerDeprecation(new \InvalidArgumentException(\sprintf(
+                'Tokens should be a list of Token instances - newval must be a Token. This will be enforced in version %d.0.',
+                Application::getMajorVersion() + 1
+            )));
```

Comment:
I guess it already fails when you try to use other values because of the `registerFoundToken` call below. `registerFoundToken` has a native `Token` type hint.

Reply:
yeah, but crash is super ugly when coming from nested function and not part of visible contract for this method (and can change any moment when internals of method change)

Reply:
tbh was wondering about same. not sure. especially as `registerFoundToken` is called conditionally

Reply:
> especially as registerFoundToken is called conditionally

Hmm yes. At first I thought the `equals` condition would always fail for values other that Tokens. But actually they can succeed.

```php
$tokens[$i] = new Token(';');
$tokens[$i] = ';'; // <-- no registerFoundToken call, so probably no failure
```

Ok so the deprecation might be better to be safe.
(but I guess you can't do anything useful with values other than Token instances)